server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -888,14 +914,7 @@ private void executePipelines(
 
         // reset the reroute flag, at the start of a new pipeline execution this document hasn't been rerouted yet
         ingestDocument.resetReroute();
-        final String originalIndex = indexRequest.indices()[0];
-        final Consumer<Exception> exceptionHandler = (Exception e) -> {
-            if (shouldStoreFailure.test(originalIndex)) {
-                listener.onResponse(IngestPipelinesExecutionResult.failAndStoreFor(originalIndex, e));
-            } else {
-                listener.onFailure(e);
-            }
-        };
+        final String originalIndex = indexRequest.index();
```

Comment:
https://github.com/elastic/elasticsearch/pull/108279#discussion_r1681141168

I added `IngestFailureStoreMetricsIT.java#testRerouteSuccessfulCorrectName` to verify that rerouting works correctly. Or did you have something else in mind?

Reply:
Fixed in [fd644eb](https://github.com/elastic/elasticsearch/pull/108279/commits/fd644eb5e29bf3b4aacd80effe1a4f13ce28b8d3)

Reply:
There's a few scenarios where I think this refactoring could still hit snags. Let's assume we have a pipeline that uses the set processor to update the `_index` field on the document to a different data stream. 

If the document later fails processing inside that pipeline it looks like the failure will be attributed to the original pipeline still since the metadata isn't updated on the document.

However, the metadata is eventually updated on the original document request. We then carry out a few more checks:

1. A document that has a self reference would fail ingestion because it is not serializable.
2. A document that has been passed to a pipeline that it has already visited during ingestion will fail to avoid infinite looping.
3. A document that has its index name changed by its final pipeline will fail because this violates an invariant of final pipelines.

In all these cases, it looks like the failure will be attributed to the new data stream that was given in the set pipeline, even though the error in the processing did not occur there.

Reply:
Sorry my comment `Fixed in [fd644eb](https://github.com/elastic/elasticsearch/pull/108279/commits/fd644eb5e29bf3b4aacd80effe1a4f13ce28b8d3)` was intended for our other discussion.

About your points here, I was under the impression that we wanted to use the destination data stream in/after reroutes? With the reasoning that if a user reroutes a document to a different data stream (and it failds), the Observability team doesn't want to be paged on it.

Reply:
We do indeed! But the reroute operation isn't actually completed until the pipeline execution it was called from is complete. That execution is only completed after those above validations occur. If those validations fail, we want to attribute the failure to the data stream we were writing to when the pipeline execution began. This is why the old code had an error handler that captured the original index within it, for a stable handle to that info.

Reply:
> if a user reroutes a document to a different data stream (and it failds), the Observability team doesn't want to be paged on it.

I'm not familiar with any plans to do paging in this way. That said, I don't think that's a reasonable approach to building the alert: If the user does something that breaks processing in a custom pipeline and doesn't reroute the document it still will have a failure showing up in the original data stream. 

If we do move forward with building those kinds of alerts, I would rather us attribute the document to the correct data stream and attempt to limit alerting based on the pipeline name that the failure occurs in.

Reply:
Done in [6a493c5](https://github.com/elastic/elasticsearch/pull/108279/commits/6a493c50b15bbf73ae394a64334133c187b62544)